src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs

```diff
@@ -233,6 +233,13 @@ protected virtual string GenerateUrl(string protocol, string host, VirtualPathDa
             // VirtualPathData.VirtualPath returns string.Empty instead of null.
             Debug.Assert(pathData.VirtualPath != null);
 
+            var url = FastGetUrl(protocol, host, pathData, fragment);
```

Comment:
This could use some documentation about the finding and the insight that led to the optimization.

> It's pretty common that `GenerateUrl` will be called with a null protocol, host and fragment so in that case ...